rudder-lang/src/generators/cfengine/syntax.rs

```diff
@@ -0,0 +1,402 @@
+                    .filter_map(|t| self.attributes.get(t).map(|p| (t, p)))
+                    .map(|(k, v)| {
+                        if first {
+                            first = false;
+                            format!("    \"{}\" {} => {}", promiser, k, v)
+                        } else {
+                            format!("    {:>width$} => {}", k, v, width = padding)
+                        }
+                    })
+                    .collect::<Vec<String>>()
+                    .join(",\n")
+            )
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub enum BundleType {
+    Agent,
+}
+
+impl fmt::Display for BundleType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                BundleType::Agent => "agent",
+            }
+        )
+    }
+}
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct Bundle {
+    bundle_type: BundleType,
+    name: String,
+    // Order matters!
+    parameters: Vec<String>,
+    // Promises of a given promise type need to be ordered specifically,
+    // hence the `Vec`.
+    promises: HashMap<PromiseType, Vec<Promise>>,
+}
+
+impl Bundle {
+    pub fn agent<T: Into<String>>(name: T) -> Self {
+        Self {
+            name: name.into(),
+            bundle_type: BundleType::Agent,
+            parameters: Vec::new(),
+            promises: HashMap::new(),
+        }
+    }
+
+    pub fn parameters(self, parameters: Vec<String>) -> Self {
+        Self { parameters, ..self }
+    }
+
+    pub fn promise(mut self, promise: Promise) -> Self {
+        self.add_promise(promise);
+        self
+    }
+
+    pub fn add_promise(&mut self, promise: Promise) {
+        match self.promises.get_mut(&promise.promise_type) {
+            Some(promises) => promises.push(promise),
+            None => {
+                self.promises.insert(promise.promise_type, vec![promise]);
+            }
+        }
+    }
+}
+
+impl fmt::Display for Bundle {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "bundle {} {}{} {{\n{}}}",
+            self.bundle_type,
+            self.name,
+            if self.parameters.is_empty() {
+                String::new()
+            } else {
+                format!("({})", self.parameters.join(", "))
+            },
+            NORMAL_ORDERING
+                .iter()
+                .filter_map(|t| self.promises.get(&t).map(|p| (t, p)))
+                .map(|(t, p)| format!(
+                    "  {}:\n{}\n",
+                    t,
+                    p.iter()
+                        .enumerate()
+                        .map(|(i, p)| format!("{}", p.format(i)))
+                        .collect::<Vec<String>>()
+                        .join("\n")
+                ))
+                .collect::<Vec<String>>()
+                .join("\n")
+        )
+    }
+}
+
+pub struct Technique {
```

Comment:
don't you want to keep the description?

Reply:
Description is a documentation field for directive creators, I don't think it really belongs in the generated policies.

Reply:
ok